File: src/Domain/Entities/Instrument.cs
using Domain.Enums;

namespace Domain.Entities;

public class Instrument
{
    public Guid Id { get; set; }
    public string Ticker { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string Currency { get; set; } = string.Empty;
    public InstrumentType Type { get; set; }

    // Identifiers for data provider lookups
    public string? Isin { get; set; }
    public string? Sedol { get; set; }
    public string? Cusip { get; set; }

    // Classification and metadata
    public AssetClass? AssetClass { get; set; }
    public string? Sector { get; set; }
    public string? Exchange { get; set; }
    public string? Country { get; set; }

    // Data provider configuration
    public PriceSource? PreferredDataProvider { get; set; }
    public string? DataProviderConfig { get; set; } // JSON string for provider-specific settings

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }

    // Navigation properties
    public virtual ICollection<Holding> Holdings { get; set; } = new List<Holding>();
    public virtual ICollection<Price> Prices { get; set; } = new List<Price>();
}

File: frontend/src/types/instrument.ts
// Instrument/Security types

export type InstrumentType = 'Cash' | 'Security';

export type AssetClass = 'Equity' | 'Bond' | 'Commodity' | 'Currency' | 'RealEstate' | 'Alternative' | 'Cash';

export type PriceSource = 'Manual' | 'AlphaVantage' | 'Finnhub' | 'YahooFinance';

export interface InstrumentDto {
  id: string;
  ticker: string;
  name: string;
  currency: string;
  type: InstrumentType;
  isin?: string | null;
  sedol?: string | null;
  cusip?: string | null;
  assetClass?: AssetClass | null;
  sector?: string | null;
  exchange?: string | null;
  country?: string | null;
  preferredDataProvider?: PriceSource | null;
  dataProviderConfig?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreateInstrumentRequest {
  ticker: string;
  name: string;
  currency: string;
  type: InstrumentType;
  isin?: string | null;
  sedol?: string | null;
  cusip?: string | null;
  assetClass?: AssetClass | null;
  sector?: string | null;
  exchange?: string | null;
  country?: string | null;
  preferredDataProvider?: PriceSource | null;
  dataProviderConfig?: string | null;
}

export interface UpdateInstrumentRequest {
  ticker: string;
  name: string;
  currency: string;
  type: InstrumentType;
  isin?: string | null;
  sedol?: string | null;
  cusip?: string | null;
  assetClass?: AssetClass | null;
  sector?: string | null;
  exchange?: string | null;
  country?: string | null;
  preferredDataProvider?: PriceSource | null;
  dataProviderConfig?: string | null;
}
